src/eddymotion/model/base.py

```diff
@@ -86,51 +94,18 @@ class BaseModel:
     __slots__ = (
         "_model",
         "_mask",
-        "_S0",
-        "_b_max",
         "_models",
         "_datashape",
     )
     _modelargs = ()
 
-    def __init__(self, gtab, S0=None, mask=None, b_max=None, **kwargs):
+    def __init__(self, mask=None, **kwargs):
         """Base initialization."""
 
-        # Setup B0 map
-        self._S0 = None
-        if S0 is not None:
-            self._S0 = np.clip(
-                S0.astype("float32") / S0.max(),
-                a_min=1e-5,
-                a_max=1.0,
-            )
+        self._model = None
 
         # Setup brain mask
         self._mask = mask
-        if mask is None and S0 is not None:
-            self._mask = self._S0 > np.percentile(self._S0, 35)
-
-        # Cap b-values, if requested
-        self._b_max = None
-        if b_max and b_max > 1000:
-            # Saturate b-values at b_max, since signal stops dropping
-            gtab[-1, gtab[-1] > b_max] = b_max
-            # A possibly good alternative is completely remove very high b-values
-            # bval_mask = gtab[-1] < b_max
-            # data = data[..., bval_mask]
-            # gtab = gtab[:, bval_mask]
-            self._b_max = b_max
-
-        kwargs = {k: v for k, v in kwargs.items() if k in self._modelargs}
-
-        model_str = getattr(self, "_model_class", None)
-        if not model_str:
-            raise TypeError("No model defined")
```

Comment:
@oesteban Some tests are now failing because the `_model_class` is `None` in this base class, and I am not setting any particular value in the derived classes. What is this property supposed to contain?

e.g.
https://app.circleci.com/pipelines/github/nipreps/eddymotion/1070/workflows/c9747d35-0cb1-49d9-963f-207d20887ce8/jobs/1038

Reply:
I am confused about the use of this. I now see that `_model_class` and `_modelargs` are properties of the wrappers `DTIModel` and `DKIModel`. This adds to this docstring https://github.com/nipreps/eddymotion/pull/176/files#diff-a875f501910044a7d95658fb83740e2c5c6c1693e7e6808703d282441db82be8L79

If I say here
```
kwargs = {k: v for k, v in kwargs.items() if k in self._modelargs}
from importlib import import_module
model_str = "eddymotion.model.AverageDWModel"
module_name, class_name = model_str.rsplit(".", 1)
my_model = getattr(import_module(module_name), class_name)(**kwargs)
```

and leaving aside that `AverageDWModel` requires a `gtab` (as it inherits from `BaseDWIModel`; setting it to `None` in its `init` would make it), the above statement produces a recursive call, since instantiating `AverageDWModel` calls the superclass `init` method.

So I am not following what was intended with this block.

Also, I am not sure what we want to do here with the DTI and DKI wrappers either.

@oesteban Can you please clarify this?

Reply:
> and leaving aside that AverageDWModel requires a gtab (as it inherits from BaseDWIModel;

The other two tests fail because of this reason.

Reply:
I'll have a look ASAP - sorry for my slow turnaround

Reply:
Sorry for the delayed response. I can now answer this question. I'm sorry for this particular case—you are prey to an undocumented feature.

The idea was that models can be fit in two ways:
- Pure leave-one-out fashion: at every iteration of the Estimator, a fully-fledged model is fit without the particular index/orientation. This is typically very slow.
- Single model: the model is fit on all the data, and each iteration produces the left-out index. These are enable by adding the prefix `Full` to the model name.

This is implemented in the Estimator, under the understanding that the model is the same, what changes is how you use it.
https://github.com/nipreps/eddymotion/blob/ce8de17b63ecce46b132d6a14d3c3ab34de3cca4/src/eddymotion/estimator.py#L117-L135

ATM I cannot comment on why this has some effect on the model itself so a `_model_class` is necessary, I can't recall the reason. I bet it is just to inform the estimator that fit should not be called every time (which probably should be handled here!)

That said, let's take the average model for example. When instantiated as `FullAverage`, then it is fit only once before entering the iterator loop of the estimator. If not, at every iteration an average without the particular direction will be calculated in the fit call.

Reply:
Okay, I'll leave my above comment because it explains something useful --- but it is totally unrelated to @jhlegarreta's question. Apologies for the confusion.

After working on the PR and re-reading the code, I understand that `_model_class` and `_modelargs` enable using DIPY models without much overhead (see DKI and DTI at the end).